pkg/storage/stores/indexshipper/compactor/compactor.go

```diff
@@ -226,53 +234,76 @@ func NewCompactor(cfg Config, objectClient client.ObjectClient, schemaConfig con
 	compactor.subservicesWatcher = services.NewFailureWatcher()
 	compactor.subservicesWatcher.WatchManager(compactor.subservices)
 
-	if err := compactor.init(objectClient, schemaConfig, limits, r); err != nil {
+	if err := compactor.init(objectStoreClients, schemaConfig, limits, r); err != nil {
 		return nil, err
 	}
 
 	compactor.Service = services.NewBasicService(compactor.starting, compactor.loop, compactor.stopping)
 	return compactor, nil
 }
 
-func (c *Compactor) init(objectClient client.ObjectClient, schemaConfig config.SchemaConfig, limits *validation.Overrides, r prometheus.Registerer) error {
+func (c *Compactor) init(objectStoreClients map[string]client.ObjectClient, schemaConfig config.SchemaConfig, limits *validation.Overrides, r prometheus.Registerer) error {
 	err := chunk_util.EnsureDirectory(c.cfg.WorkingDirectory)
 	if err != nil {
 		return err
 	}
-	c.indexStorageClient = shipper_storage.NewIndexStorageClient(objectClient, c.cfg.SharedStoreKeyPrefix)
-	c.metrics = newMetrics(r)
 
 	if c.cfg.RetentionEnabled {
-		var encoder client.KeyEncoder
-		if _, ok := objectClient.(*local.FSObjectClient); ok {
-			encoder = client.FSEncoder
+		deleteRequestStore := c.cfg.DeleteRequestStore
+		// if -boltdb.shipper.compactor.shared-store is set, use it instead to ensure backward compatibility.
+		if c.cfg.SharedStoreType != "" {
+			deleteRequestStore = c.cfg.SharedStoreType
 		}
 
-		chunkClient := client.NewClient(objectClient, encoder, schemaConfig)
+		objectClient, ok := objectStoreClients[deleteRequestStore]
+		if !ok {
+			return fmt.Errorf("failed to init delete store. Object client not found for %s", c.cfg.DeleteRequestStore)
+		}
 
-		retentionWorkDir := filepath.Join(c.cfg.WorkingDirectory, "retention")
-		c.sweeper, err = retention.NewSweeper(retentionWorkDir, chunkClient, c.cfg.RetentionDeleteWorkCount, c.cfg.RetentionDeleteDelay, r)
-		if err != nil {
-			return err
+		if err := c.initDeletes(objectClient, r, limits); err != nil {
+			return fmt.Errorf("failed to init delete store: %w", err)
 		}
 
-		if err := c.initDeletes(r, limits); err != nil {
-			return err
+		if err := retention.MigrateMarkers(filepath.Join(c.cfg.WorkingDirectory, "retention"), deleteRequestStore); err != nil {
```

Comment:
deletion markers might not always belong to objects in the delete requests store. There is a high chance when someone wants to start using a new bucket, they would change the bucket for delete requests store as well. In that case we will try deleting objects from new bucket here while the objects were stored in older bucket. Will think more and see what we can do here.